datamodel/include/datamodel/math/MatrixColumnMajorAdaptor.hpp

```diff
@@ -1,25 +1,59 @@
 #ifndef RCBE_ENGINE_MATRIXCOLUMNMAJORADAPTOR_HPP
 #define RCBE_ENGINE_MATRIXCOLUMNMAJORADAPTOR_HPP
 
+#include <array>
 #include <cstdint>
+#include <iostream>
 
 #include <rcbe-engine/datamodel/math/Matrix.hpp>
 
+#ifdef RCBE_GLM_INTEGRATION
+#include <glm/matrix.hpp>
+#endif
+
+#define RCBE_DEBUG_MODE
+
 namespace rcbe::math {
+template <typename StoredValue = core::EngineScalar>
 class MatrixColumnMajorAdaptor {
 public:
-    using StorageType = std::vector<core::EngineScalar>;
+    using ValueType = StoredValue;
+    using StorageType = std::vector<ValueType>;
 
     MatrixColumnMajorAdaptor() = default;
     ~MatrixColumnMajorAdaptor() = default;
 
+#ifdef RCBE_GLM_INTEGRATION
+    template <glm::length_t r, glm::length_t c, typename T, glm::qualifier Q>
+    explicit MatrixColumnMajorAdaptor(const glm::mat<c, r, T, Q> &matrix)
+    {
+        adapted_storage_.reserve(r * c);
+        for (size_t i = 0; i < c; ++i) {
+            for (size_t j = 0; j < r; ++j) {
+                adapted_storage_.push_back(static_cast<ValueType>(matrix[i][j]));
+            }
+        }
+    }
+#endif
+
+    template <typename Value, size_t sz>
+    explicit MatrixColumnMajorAdaptor(std::array<Value, sz> storage)
```

Comment:
Kind of puzzling, the whole purpose of this class is to change the memory layout of your matrix. Basically from the case where rows are sequentially stored in the array we move to the case when columns a stored sequentially. This ctor looks strange. Is it actually needed?

Reply:
Hid one of them strange ctors under `#ifdef RCBE_DEBUG_MODE`. The other one is needed, though.